src/ui/textfield.rs

```diff
@@ -1,15 +1,63 @@
 //! A fairly simplistic text field implementation.
 
+use std::{borrow::BorrowMut, ops::Range};
+
+use copypasta::{ClipboardContext, ClipboardProvider};
 use skulpin::skia_safe::*;
 
 use crate::ui::*;
 
+/// Text field selection struct
+struct Selection {
+    pub first: usize,
+    pub second: usize,
+}
+
+impl Selection {
+    pub fn start(&self) -> usize {
+        self.first.min(self.second)
+    }
+
+    pub fn end(&self) -> usize {
+        self.first.max(self.second)
+    }
+
+    pub fn normalize(&self) -> Range<usize> {
+        self.start()..self.end()
+    }
+
+    pub fn len(&self) -> usize {
+        self.end() - self.start()
+    }
+
+    pub fn move_cursors(&mut self, position: usize) {
+        self.first = position;
+        self.second = self.first;
+    }
+
+    pub fn move_cursors_left(&mut self) {
+        if self.first > 0 {
+            self.first -= 1;
+            self.second = self.first;
+        }
+    }
+
+    pub fn move_cursors_right(&mut self) {
+        self.first += 1;
+        self.second = self.first;
+    }
+}
```

Comment:
Similarly these should be named `move_left` and `move_right`. Also, try hoisting the moving-while-Shift-is-down logic into these functions by adding a `shift: bool` parameter that will disable setting `self.second` to `self.first`.